compiler/hash-reporting/src/render.rs

```diff
@@ -111,6 +111,118 @@ impl ReportCodeBlock {
             .take(top_buffer + end_row - start_row + 1 + bottom_buffer)
     }
 
+    /// Function which compute the "true" width of a span in a source file. This
+    /// means that if the line contains any [combining marks][https://unicode.org/reports/tr15/#Normalization_Forms_Table].
+    ///
+    /// If any combining marks are encountered, it is assumed that they are not
+    /// rendered within the span, and thus they are subtracted from the
+    /// total width of the given drawn span.
+    ///
+    /// More specifically, when the rendering wants to draw a reference on a
+    /// span which will be of some "width", it will use this function to
+    /// compute the "true" width of the span, and then draw the span using
+    /// the computed width. For example, this snippet of code:
+    ///
+    /// ```
+    /// 'ä' // a + combining diaeresis'
+    /// ```
+    ///
+    /// In this snippet, the compiler lexer will emit an error on the `ä`
+    /// character because it is not normalised. The compiler will then want
+    /// to draw a span on the `'ä'` character to indicate that it is the
+    /// source of the error. However, the `'ä'` character is actually
+    /// two characters, the `a` and the combining diaeresis. So, when the
+    /// compiler wants to draw the following (snippet 1):
+    /// ```
+    /// 'ä'
+    /// ^^^
+    /// ```
+    ///
+    /// It will actually draw the following (snippet 2):
+    ///
+    /// ```
+    /// 'ä'
+    /// ^^^^
+    /// ```
+    ///
+    /// To solve this issue, we need to compute the "true" width of the span,
+    /// which is the width of the span minus the number of combining marks
+    /// in the span. So, in the above example, the "true" width of the span
+    /// is 3, and thus the compiler will draw (snippet 1) instead of
+    /// (snippet 2).
+    ///
+    /// N.B. It is assumed that the provided [RowColRange] references on the
+    /// `line`.
+    fn get_line_display_width(&self, line: &str, start: usize, end: usize) -> usize {
+        // count the number of combining marks is up to the given `end.column`.
+        let mut combining_marks = 0;
+
+        for (index, ch) in line.chars().enumerate().skip(start) {
+            // Stop counting the characters once we reach the end of the span.
+            if index == end {
+                break;
+            }
+
+            // If it is a combining mark, its not going to be rendered within the
+            // span, so we need to subtract it from the total width.
+            if unicode_normalization::char::is_combining_mark(ch) {
+                combining_marks += 1;
+            }
+        }
+
+        (end - start).saturating_sub(combining_marks)
+    }
```

Comment:
Interesting approach. Instead of the `unicode_normalization` crate, `unicode_segmentation` could be used which offers a `.graphemes()` iterator so that this function could be replaced by `.graphemes().count()`. Not necessary though, just a potential simplification.

Reply:
There is also the `unicode-width` crate, which should compute the true column width of a given Unicode string. Perhaps it is worth leaving a todo/investigate note?